Document array and math functions (#82)
sanity-io/GROQ: spec/Section 10 -- Functions.md

string::joinValidate(args):

- If the length of {args} is not 1:
  - Report an error.

### array::unique()

The `unique` function filters duplicate values from an array.

When considering uniqueness, only primitive values (numbers, strings, booleans, date-times, nulls)
must considered for equality. Non-primitive values (objects and arrays) must always be considered

> separately unique. For example, `array::unique([[1], [1]])` should return `[[1], [1]]`.

Reviewer: I don't think we need to define "primitive values" here as this behavior is already defined under https://sanity-io.github.io/GROQ/draft/#sec-Equality.

Author: Good points. Updated.

Since an implementation can choose to use hashing or similar data non-order-preserving structures
for efficiency, the order of the output is not guaranteed to be the same as the input.

array::unique(args, scope):

- Let {arrNode} be the first element of {args}.
- Let {arr} be the result of {Evaluate(arrNode, scope)}.
- If {arr} is not an array:
  - Return {null}.
- Let {output} be an empty array.
- For each element in {arr}:
  - Let {elem} be the element
  - If {elem} is not a primitive:
    - Append {elem} to {output}.
  - If {elem} is not in {output}:

>     - Append {elem} to {output}.

Reviewer: This should refer to https://sanity-io.github.io/GROQ/draft/#sec-Equality to clarify exactly what we mean by "is not in". This would also remove the special case of "is not a primitive" (since `[] != []`).

>     - Append {elem} to {output}.

Reviewer: How about something like this to clarify that the order is not guaranteed?

```suggestion
    - Add {elem} into {output} (at any position).
```

ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns 0.

math:sum(args, scope):

- Let {arrNode} be the first element of {args}.
- Let {arr} be the result of {Evaluate(strNode, scope)}.
- If {arr} is not an array, return {null}.
- Let {n} be zero.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:
    - Add {elem} to {n}.
- Return {n}.

math:sumValidate(args):

- If the length of {args} is not 1:
  - Report an error.

### math::avg()

The `avg` function computes the arithmetic mean of all numbers in an array. {null} values are
ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns {null}.

math:avg(args, scope):

- Let {arrNode} be the first element of {args}.
- Let {arr} be the result of {Evaluate(strNode, scope)}.
- If {arr} is not an array, return {null}.
- Let {n} be zero.
- Let {count} be zero.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.

>   - Otherwise:
>     - Increment {count}.

Reviewer 2: ```suggestion
  - Otherwise:
    - Increment {count}.
    - let {n} be the sum of {elem} and {n}.
```

one numeric value, it returns {null}.

math:min(args, scope):

- Let {arrNode} be the first element of {args}.
- Let {arr} be the result of {Evaluate(strNode, scope)}.
- If {arr} is not an array, return {null}.
- Let {min} be {null}.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:
    - If {min} is {null} or {elem} is lower than {min}:
      - Set {min} to {elem}.
- Return {min}.

math:minValidate(args):

- If the length of {args} is not 1:
  - Report an error.

### math::max()

The `max` function finds the largest numeric value in an array. {null} values are
ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns {null}.

math:max(args, scope):

- Let {arrNode} be the first element of {args}.
- Let {arr} be the result of {Evaluate(strNode, scope)}.
- If {arr} is not an array, return {null}.
- Let {max} be {null}.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:

>     - If {max} is {null} or {elem} is lower than {max}:

Reviewer 2: ```suggestion
    - If {max} is {null} or {elem} is more than {max}:
```

- If the length of {args} is not 2:
  - Report an error.

## Math namespace

The `math` namespace contains functions for performing mathematical operations.

### math::sum()

The `sum` function computes the sum of all numbers in an array. {null} values are
ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns 0.

math:sum(args, scope):

- Let {arrNode} be the first element of {args}.

> - Let {arr} be the result of {Evaluate(strNode, scope)}.

Reviewer: ```suggestion
- Let {arr} be the result of {Evaluate(arrNode, scope)}.
```

- If {arr} is not an array, return {null}.
- Let {n} be zero.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:
    - Add {elem} to {n}.
- Return {n}.

math:sumValidate(args):

- If the length of {args} is not 1:
  - Report an error.

### math::avg()

The `avg` function computes the arithmetic mean of all numbers in an array. {null} values are
ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns {null}.

math:avg(args, scope):

- Let {arrNode} be the first element of {args}.

> - Let {arr} be the result of {Evaluate(strNode, scope)}.

Reviewer: ```suggestion
- Let {arr} be the result of {Evaluate(arrNode, scope)}.
```

- If {arr} is not an array, return {null}.
- Let {n} be zero.
- Let {count} be zero.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:
    - Increment {count}.
    - Add {elem} to {n}.
- If {count} is zero:
  - Return {null}.
- Return {n} divided by the {count}.

math:avgValidate(args):

- If the length of {args} is not 1:
  - Report an error.

### math::min()

The `min` function finds the smallest numeric value in an array. {null} values are
ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns {null}.

math:min(args, scope):

- Let {arrNode} be the first element of {args}.
- Let {arr} be the result of {Evaluate(strNode, scope)}.
- If {arr} is not an array, return {null}.
- Let {min} be {null}.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:

>     - If {min} is {null} or {elem} is lower than {min}:

Reviewer: ```suggestion
    - If {min} is {null} or {PartialCompare(elem, min)} is {Less}:
```

      - Set {min} to {elem}.
- Return {min}.

math:minValidate(args):

- If the length of {args} is not 1:
  - Report an error.

### math::max()

The `max` function finds the largest numeric value in an array. {null} values are
ignored, but non-numbers cause the function to return {null}. If the array does not contain at least
one numeric value, it returns {null}.

math:max(args, scope):

- Let {arrNode} be the first element of {args}.

> - Let {arr} be the result of {Evaluate(strNode, scope)}.

Reviewer: ```suggestion
- Let {arr} be the result of {Evaluate(arrNode, scope)}.
```

- If {arr} is not an array, return {null}.
- Let {max} be {null}.
- For each element {elem} in {arr}:
  - If {elem} is null:
    - Ignore it.
  - If {elem} is not a number:
    - Return {null}.
  - Otherwise:

>     - If {max} is {null} or {elem} is higher than {max}:

Reviewer: ```suggestion
    - If {max} is {null} or {PartialCompare(elem, max)} is {Greater}:
```
